File: lib/test/structuredfield/parseString.test.ts
import assert from 'node:assert';
import test from 'node:test';
import { parseString } from '../../src/structuredfield/parse/parseString.js';

test('parseString', () => {
	assert.deepStrictEqual(parseString(`"asdf"`), { value: `asdf`, src: `` });
	assert.deepStrictEqual(parseString(`"!#[]"`), { value: `!#[]`, src: `` });
	assert.deepStrictEqual(parseString(`"a""`), { value: `a`, src: `"` });
	assert.deepStrictEqual(parseString(`"a\\""`), { value: `a"`, src: `` });
	assert.deepStrictEqual(parseString(`"a\\\\c"`), { value: `a\\c`, src: `` });
	assert.throws(() => parseString(`"a\\"`), /failed to parse ""a\\"" as String/);
	assert.throws(() => parseString(`"\\a"`), /failed to parse ""\\a"" as String/);
	assert.throws(() => parseString(``), /failed to parse "" as String/);
});

File: python/src/test/structuredfield/parseString_test.py
import unittest

from ...structuredfield.parse.ParsedValue import ParsedValue
from ...structuredfield.parse.parseString import parseString
from ...structuredfield.SfException import SfException


class TestParseToken(unittest.TestCase):
    def test_parseString(self):
        self.assertEqual(parseString('"asdf"'), ParsedValue("asdf", ""))
        self.assertEqual(parseString('"!#[]"'), ParsedValue("!#[]", ""))
        self.assertEqual(parseString('"a""'), ParsedValue("a", '"'))
        self.assertEqual(parseString('"a\\""'), ParsedValue('a"', ""))
        self.assertEqual(parseString('"a\\\\c"'), ParsedValue("a\\c", ""))

        with self.assertRaises(SfException) as context:
            parseString('"a\\"')
        self.assertEqual(context.exception.message, 'failed to parse ""a\\"" as String')

        with self.assertRaises(SfException) as context:
            parseString('"\\a"')
        self.assertEqual(context.exception.message, 'failed to parse ""\\a"" as String')

        with self.assertRaises(SfException) as context:
            parseString("")
        self.assertEqual(context.exception.message, 'failed to parse "" as String')
